### PR TITLE
[Darwin][Driver][clang] Prioritise command line args over `DEFAULT_SYSROOT`

### PR DESCRIPTION
**Pull request by carlocab**
If a toolchain is configured with `DEFAULT_SYSROOT`, then this could
result in an unintended value for `-syslibroot` being passed to the
linker if the user manually sets `-isysroot` or `SDKROOT`.

Let's fix this by prioritising command line flags when determining
`-syslibroot` before checking `getSysRoot`.

Downstream bug report: https://github.com/Homebrew/homebrew-core/issues/197277

Co-authored-by: Bo Anderson <mail@boanderson.me>

### CODE REVIEW

**Comment by jyknight** [comment]
This is the intended/expected behavior.

On every other platform, `-isysroot` is supposed to be an override setting the sysroot only for the header files, and _never_ overriding the linker sysroot. The `--sysroot` flag, on the other hand, sets the value used for both libraries and headers (unless the header sysroot is set via `-isysroot`). As such, there's almost never a reason to pass `-isysroot` -- the right flag is `--sysroot` 99% of the time.

Except, on Darwin platforms, for some reason that's probably lost to history, the `-isysroot` is used almost universally instead of `--sysroot`, and a platform special-case was added to cause `-isysroot` to affect both the header _and_ library sysroots -- unless the `--sysroot` flag is also set. This is weird and inconsistent, but it's been that way forever. The behavior was implemented very early on in Clang's development, because the pre-existing Darwin GCC port did that. 

However, notably, even on Darwin, if you pass _both_ flags, you still get the intended behavior of `-isysroot` affecting headers, and `--sysroot` affecting libraries. Your change would break that.

**Comment by carlocab** [comment]
> However, notably, even on Darwin, if you pass _both_ flags, you still get the intended behavior of `-isysroot` affecting headers, and `--sysroot` affecting libraries. Your change would break that.

What's the use-case for this?

Because for most users, you almost certainly don't want to mix headers and libraries of different versions, and keeping the existing behavior is essentially a big footgun when trying to build a useful toolchain on Darwin.

At Homebrew, we used to configure the toolchain with `DEFAULT_SYSROOT` but have recently switched that to setting the `--sysroot` flag in a Clang `.cfg` file.[^1] In either case, if you set `SDKROOT` in your environment or try to run our `clang` using Apple's `xcrun`, this becomes equivalent to setting `-isysroot` and you most likely will end up with headers that mismatch the libraries you try to link to.

It seems to me that getting rid of a footgun that can impact many users (our toolchain is installed [a few hundred thousand times a year](https://formulae.brew.sh/formula/llvm#default), see also the [downstream bug report](https://github.com/Homebrew/homebrew-core/issues/197277) I initially linked) might be worth breaking what currently seems to be a niche use-case of using headers and libraries of different versions. If you really wanted to do this, you can set `-Wl,-syslibroot`.

Of course, I'm open to alternatives that don't involve breaking anyone.

[^1]: We do this so that the toolchain can find the sysroot out of the box, when it wouldn't be able to otherwise.

**Comment by MaskRay** [comment]
(I am not familiar with macOS and I'll defer the decision to Apple folks.
I agree with James that this PR might break users.

You probably want a different way to handle default sysroot in homebrew.
)

**Comment by cyndyishida** [comment]
> > However, notably, even on Darwin, if you pass _both_ flags, you still get the intended behavior of `-isysroot` affecting headers, and `--sysroot` affecting libraries. Your change would break that.
> 
> What's the use-case for this?
> 
> Because for most users, you almost certainly don't want to mix headers and libraries of different versions, and keeping the existing behavior is essentially a big footgun when trying to build a useful toolchain on Darwin.
> 
> At Homebrew, we used to configure the toolchain with `DEFAULT_SYSROOT` but have recently switched that to setting the `--sysroot` flag in a Clang `.cfg` file.[1](#user-content-fn-1-21a5728b47e790352033de78c44bc2e4) In either case, if you set `SDKROOT` in your environment or try to run our `clang` using Apple's `xcrun`, this becomes equivalent to setting `-isysroot` and you most likely will end up with headers that mismatch the libraries you try to link to.

The expected flow on Apple platforms is to only pass an `isysroot` argument whether it's inherited from `xcrun -sdk <sdk version> clang` or passed explicitly. Could homebrew instead only pass `isysroot` for Darwin targets? Or check that `sysroot` and `isysroot` inputs are the same when creating an invocation.  It's unclear to me how homebrew gets into a situation where there are conflicting sdks passed for `sysroot` and `isysroot` when you effectively want to ignore whatever is passed to `sysroot`.

**Comment by cachemeifyoucan** [comment]
I think the correct solution for Homebrew is to never use `--sysroot` and switch to `-isysroot` since the former is almost never used on Darwin platforms.

**Comment by Bo98** [comment]
> The expected flow on Apple platforms is to only pass an `isysroot` argument whether it's inherited from `xcrun -sdk <sdk version> clang` or passed explicitly. Could homebrew instead only pass `isysroot` for Darwin targets? Or check that `sysroot` and `isysroot` inputs are the same when creating an invocation. It's unclear to me how homebrew gets into a situation where there are conflicting sdks passed for `sysroot` and `isysroot` when you effectively want to ignore whatever is passed to `sysroot`.

In the past it was as simple as it happens when setting `DEFAULT_SYSROOT`. If only `isysroot` is supported on macOS, does this mean `DEFAULT_SYSROOT` should be deprecated on Apple platforms?

**Comment by cyndyishida** [comment]
> The main concern is it happens when setting `DEFAULT_SYSROOT`. If only `isysroot` is supported on macOS, does this mean `DEFAULT_SYSROOT` should be deprecated on Apple platforms?

I don't think it's deprecated in the sense we have plans to drop support for it. The problem here seems more like when those values conflict what should be expected?

Today it seems expected that `sysroot` will be used for library search but not header search, which also matches the other platforms. I don't know if this adds confusion or helps unblock, but clang on Darwin also respects the env var `SDKROOT` which translates to `isysroot`.

**Comment by Bo98** [comment]
> I don't think it's deprecated in the sense we have plans to drop support for it. The problem here seems more like when those values conflict what should be expected?
>
> Today it seems expected that `sysroot` will be used for library search but not header search, which also matches the other platforms. I don't know if this adds confusion or helps unblock, but clang on Darwin also respects the env var `SDKROOT` which translates to `isysroot`.

I do agree with this and this does make sense. The problem has mostly come from two expectations:
* `DEFAULT_SYSROOT` was expected by some users to support setting a default SDK. Maybe this wasn't actually intended, but it came common enough knowledge to eventually be explicitly recommended in parts of LLVM documentation: https://github.com/llvm/llvm-project/blob/ee9e7867178bee16cc0976a83a8303c99c2f5326/flang/docs/GettingStarted.md?plain=1#L91-L93.
* `-isysroot` generally seems to preferred on macOS over `--sysroot`, despite that not being the case on other platforms.

However those two points are incompatible. And `DEFAULT_SYSROOT` was, until the introduction of config files recently, the only way to set a default as the Apple's SDK driver is closed source. We used `DEFAULT_SYSROOT` since the SDK changes in 10.14, but config files are becoming a possible alternative with #111387.

---

This probably crosses into #38193, but to paint the wider picture since I agree the problem here is more complicated than a prioritisation problem, it's generally expected from various build scripts that something like this:

```clang
#include <stdio.h>

int main() {
   printf("Hello, World!");
   return 0;
}
```

will just work out of the box without needing extra flags or to pipe it through other commands. It works out of the box on Linux and likely most other platforms but does not on macOS.

**Comment by cachemeifyoucan** [comment]
> DEFAULT_SYSROOT was expected by some users to support setting a default SDK. Maybe this wasn't actually intended, but it came common enough knowledge to eventually be explicitly recommended in parts of LLVM documentation:

That is flang document, which is not endorsed by platform vendor in anyway. You can use DEFAULT_SYSROOT for local builds (or package manager builds for local host only) but never for products that are built for distribution on macOS. There is no standard location for macOS SDK, and it can be anywhere you want. You have to specify `-isysroot` or set env `SDKROOT` (which is what /usr/bin/clang does, it is a shim that finds the SDK location, then set SDKROOT). DEFAULT_SYSROOT can't fix any problem for distribution.

All Darwin platforms use a different model from your standard linux distribution and all SDKs are backwards compatible to older OS versions, which allows you to build software for older OS on latest version. The compilation model is closer to a cross-compilation in linux world. 

The fundamental problem you hit is that you pass a secret `--sysroot` option (not visible from driver command) that doesn't match the `-isysroot`. If you intended to make this less a problem for unsuspecting users, I would change to pass `-isysroot` on Darwin, instead of using `--sysroot` for DEFAULT_SYSROOT. That should be safe enough to not disrupt any users.

**Comment by Bo98** [comment]
> There is no standard location for macOS SDK, and it can be anywhere you want. DEFAULT_SYSROOT can't fix any problem for distribution.

To be clear: the build I talked about there required the CLT where the SDK location is standardised, so this isn't entirely true for the use case I had. But yes a general LLVM solution would require something more like `xcrun`.

The major-fixed `MacOSX.sdk` (e.g. `MacOSX15.sdk`) distributed with the CLT was in fact something we requested from Apple a few years back and was implemented for the precise use case where downstream tools (mostly GCC but it ended up being used for LLVM too) needed a fixed location. (I don't have the `rdar` number on hand as I don't have the conversation saved but hopefully you get what I mean here)

> all SDKs are backwards compatible to older OS versions, which allows you to build software for older OS on latest version

This is mostly true but there are notable exceptions such as:
* `libcurl` does not have OS availability APIs and it is known that using `libcurl` headers can lead to runtime crashes on older OS
* `unguarded-availability-new` is a warning by default and not all configure scripts set `-Werror` for it so will not detect feature availability properly

And we have seen these break real world applications, even high profile ones like `git`.

This doesn't really change anything here though. Distributing `clang` in a way that doesn't need extra flags is still an expectation the vast majority of downstream build scripts expect, regardless what SDK that is.

>  If you intended to make this less a problem for unsuspecting users, I would change to pass -isysroot on Darwin, instead of using `--sysroot` for `DEFAULT_SYSROOT`. That should be safe enough to not disrupt any users.

`DEFAULT_SYSROOT` is a LLVM feature. It cannot be made to use `-isysroot`.

Migrating to config files does give us the option to use `-isysroot` instead and we will do that. The v1 initial port to config files was simply a 1:1 behavioural match to the previous `DEFAULT_SYSROOT` behaviour. We've been contributing patches upstream to make config files work better for this use case (such as the separate Darwin PR linked above) and getting feedback from users from edge cases where config files still aren't being applied correctly such as `clang-tidy`.

**Comment by cachemeifyoucan** [comment]
Ah, I see. I think `--sysroot` is not even involved when you use `DEFAULT_SYSROOT`, it is injected here:
https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/Driver.cpp#L205

Maybe instead of, the logic in the patch should be setting `-syslibroot` for linker in following order:
* `--sysroot`
* `-isysroot`
* `C.getSysRoot()`, which is basically DEFAULT_SYSROOT as other cases is handled in `--sysroot`

I don't think DEFAULT_SYSROOT is a common setting, and if it is set, users are probably expect to overwrite using `-isysroot` or `--sysroot`.

**Comment by Bo98** [comment]
That proposal makes sense to me.

**Comment by carlocab** [comment]
> Maybe instead, the logic in the patch should be setting `-syslibroot` for linker in following order:
> 
> * `--sysroot`
> * `-isysroot`
> * `C.getSysRoot()`, which is basically DEFAULT_SYSROOT as other cases is handled in `--sysroot`

Thanks, pushed a change that I think should implement this. Haven't gotten around to testing it yet, I will this afternoon.

**Comment by carlocab** [comment]
> It's unclear to me how homebrew gets into a situation where there are conflicting sdks passed for `sysroot` and `isysroot` when you effectively want to ignore whatever is passed to `sysroot`.

Maybe this has already been answered, but just in case. You get conflicting SDKs when:

- You build a toolchain that is configured with `DEFAULT_SYSROOT` (or, you use a Clang configuration file that uses `--sysroot=`, which mostly mirrors the effects of setting `DEFAULT_SYSROOT`)
- Your user uses your toolchain and passes `-isysroot` to a different SDK (or, equivalently, sets `SDKROOT` pointing to a different SDK)

In each of the four possible scenarios above (two possibilities in each bullet point), you end up with `clang` using one SDK for headers (coming from `-isysroot`/`SDKROOT`) and a different one for libraries (coming from `DEFAULT_SYSROOT`/`--sysroot`).

> Hm, I had a thought which might meet the needs here, and not add additional confusion: we could have the Darwin SDKROOT environment variable override both `-isysroot` and `--sysroot`. Would that help the original issue?

This would help too, actually. I'll look into this. Thanks.

> I think the correct solution for Homebrew is to never use `--sysroot` and switch to `-isysroot` since the former is almost never used on Darwin platforms.

I'm looking at this in https://github.com/Homebrew/homebrew-core/pull/200047. However, based on [this comment](https://github.com/Homebrew/homebrew-core/pull/200047#issuecomment-2517969578), it doesn't quite solve the downstream bug report entirely. But that's something I'll have to look into.

Us switching to using `-isysroot` instead doesn't solve the problem of (other) users who configure a toolchain with `DEFAULT_SYSROOT`, but maybe the changes I've made here that you suggested at https://github.com/llvm/llvm-project/pull/115993#issuecomment-2518537500 will be enough for that.

**Comment by cachemeifyoucan** [comment]
Current patch breaks test:
```
// We pass --sysroot="" to defeat any -DDEFAULT_SYSROOT parameter.
```
I think that just a pure hack in the test case, we can just update the test. @cyndyishida what do you think about the current approach?

**Comment by carlocab** [comment]
We could probably do something like
```diff
diff --git a/clang/lib/Driver/ToolChains/Darwin.cpp b/clang/lib/Driver/ToolChains/Darwin.cpp
index cdb6d21a0148..f3769ffcb4ae 100644
--- a/clang/lib/Driver/ToolChains/Darwin.cpp
+++ b/clang/lib/Driver/ToolChains/Darwin.cpp
@@ -432,7 +432,7 @@ void darwin::Linker::AddLinkArgs(Compilation &C, const ArgList &Args,
   // --isysroot as the syslibroot.
   // We check `OPT__sysroot_EQ` directly instead of `getSysRoot` to make sure we
   // prioritise command line arguments over configuration of `DEFAULT_SYSROOT`.
-  if (const Arg *A = Args.getLastArg(options::OPT__sysroot_EQ)) {
+  if (const Arg *A = Args.getLastArg(options::OPT__sysroot_EQ); A && strcmp(A->getValue(), "")) {
     CmdArgs.push_back("-syslibroot");
     CmdArgs.push_back(A->getValue());
   } else if (const Arg *A = Args.getLastArg(options::OPT_isysroot)) {
```
to retain the current behaviour and avoid breaking the test. Not sure it's worth going to the trouble though.

**Comment by cyndyishida** [comment]
> Current patch breaks test:
> 
> ```
> // We pass --sysroot="" to defeat any -DDEFAULT_SYSROOT parameter.
> ```
> 
> I think that just a pure hack in the test case, we can just update the test. @cyndyishida what do you think about the current approach?

SGTM too. I don't see wide usage of it on my end. It looks like that part of test itself was added to workaround a toolchain build configuration issue, where `DEFAULT_SYSROOT` is actually used. I think fixing the test to adapt to the changes in the PR is the right move.  More details of that in: https://github.com/llvm/llvm-project/issues/94284 

Changing the test itself (to e.g. locally set that env var) is a good regression test that should be added for this pr anyway.

**Comment by carlocab** [comment]
Dropped the `--sysroot=""` flag from the test.

> Changing the test itself (to e.g. locally set that env var) is a good regression test that should be added for this pr anyway.

Sorry, which env var are you referring to?

**Comment by cyndyishida** [comment]
> Sorry, which env var are you referring to?

I'm suggesting an explicit test case that checks that `DEFAULT_SYSROOT` is set and behaves as expected. It's ignored if  sysroot or isysroot is passed, it's used when they're missing.

**Comment by carlocab** [comment]
> > Sorry, which env var are you referring to?
> 
> I'm suggesting an explicit test case that checks that `DEFAULT_SYSROOT` is set and behaves as expected. It's ignored if sysroot or isysroot is passed, it's used when they're missing.

`DEFAULT_SYSROOT` is not an environment variable. It is a CMake variable/macro that must be set at configure-time:

https://github.com/llvm/llvm-project/blob/aac000a01ba8c0ed15fec3c198fce9f13fc83725/clang/CMakeLists.txt#L208-L209

https://github.com/llvm/llvm-project/blob/aac000a01ba8c0ed15fec3c198fce9f13fc83725/clang/include/clang/Config/config.h.cmake#L48-L49

I don't think `lit` knows how to tell whether `DEFAULT_SYSROOT` is set. We could maybe teach it to, but given that that intent of #94284 is to deprecate it (and I think all usage of it in existing builders has been removed), that seems like the wrong direction.

**Review by cachemeifyoucan — APPROVED** [review]
LGTM

**Comment by carlocab** [comment]
> The PR description (first comment), which will be used as the default commit message, should be fixed.

Thanks for the reminder, fixed.

I plan to do a few more local tests with `DEFAULT_SYSROOT` enabled before merging (since this isn't tested in CI).

**Comment by carlocab** [comment]
Tested locally; seems to be working as intended.

**Pull request merged**
